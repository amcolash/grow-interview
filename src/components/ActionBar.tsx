import React from 'react';
import { style } from 'typestyle';

const barStyle = style({
  display: 'flex',
  width: '100%',
  height: '96px',
  padding: '16px',
  justifyContent: 'center',
  alignItems: 'center',
  gap: '16px',

  borderRadius: '100px',
  background: 'var(--neutral-000, #FFF)',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function ActionBar() {
  return <div className={barStyle}></div>;
}
