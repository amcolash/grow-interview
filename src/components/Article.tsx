import React from 'react';
import { style } from 'typestyle';

interface ArticleProps {
  index: number;
  title: string;
  views: number;
}

const articleStyle = style({
  display: 'flex',
  padding: '24px',
  alignItems: 'center',
  gap: '20px',
  alignSelf: 'stretch',
  fontSize: '16px',
  lineHeight: '24px',

  borderRadius: '12px',
  border: '1px solid var(--neutral-300, #E7EAEB)',
  background: 'var(--white, #FFF)',

  $nest: {
    '& .index, & .views': {
      color: 'var(--neutral-500, #A7AAAB)',
    },
    '.title': {
      flex: 1,
    },
    '.views': {
      fontSize: '14px',
      fontFamily: 'Poppins',
      lineHeight: '16px',
    },
  },
});

export function Article(props: ArticleProps) {
  return (
    <article className={articleStyle}>
      <div className="index">{props.index}</div>
      <div className="title">{props.title}</div>
      <div className="views">{props.views.toLocaleString('en-US')} views</div>
    </article>
  );
}
