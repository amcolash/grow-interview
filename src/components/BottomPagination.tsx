import React from 'react';
import ReactPaginate from 'react-paginate';
import { style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface BottomPaginationProps {
  totalPages: number;
  currentPage: number;
  onPageChange: (selectedItem: { selected: number }) => void;
}

const pageListStyle = style({
  display: 'flex',
  justifyContent: 'center',
  flexWrap: 'wrap',
  gap: '8px',
  listStyle: 'none',
  margin: 0,
  padding: 0,

  $nest: {
    '& a': {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '40px',
      height: '40px',
      padding: '10px',
      color: 'var(--neutral-900, #05090D)',
      fontFamily: 'Poppins',
    },
    '& li': {
      border: '1px solid var(--neutral-400, #D4D8D9)',
      background: 'var(--neutral-000, #FFF)',
      borderRadius: '100px',
      userSelect: 'none',

      $nest: {
        '&.disabled': {
          background: 'var(--neutral-400, #D4D8D9)',
        },
        '&.disabled > a': {
          color: 'var(--neutral-600, #737680)',
        },
        '&:not(.disabled,.selected):hover': {
          background: 'var(--neutral-300, #e7eaeb)',
        },
        '&.selected': {
          background: 'var(--avocado-300, #E0E9CB)',
        },
        '&.selected > a': {
          color: 'var(--brand-green-500, #025B4B)',
        },
      },
    },
  },
});

export function BottomPagination(props: BottomPaginationProps) {
  return (
    <ReactPaginate
      className={pageListStyle}
      breakLabel="..."
      nextLabel={<ChevronRight />}
      onPageChange={props.onPageChange}
      pageCount={props.totalPages}
      previousLabel={<ChevronLeft />}
      renderOnZeroPageCount={null}
    />
  );
}
