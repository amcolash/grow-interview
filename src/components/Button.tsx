import React, { ButtonHTMLAttributes } from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const buttonStyle = style(
  {
    display: 'flex',
    width: '160px',
    padding: '12px 24px',
    justifyContent: 'center',
    alignItems: 'center',
    gap: '8px',
    flexShrink: '0',
    alignSelf: 'stretch',
    borderRadius: '100px',
    background: 'var(--brand-green-500, #025b4b)',
    border: 'none',
    color: 'var(--neutral-000, #fff)',
    textAlign: 'center',
    fontSize: '16px',
    fontFamily: 'Poppins',
    fontWeight: '500',
    lineHeight: '24px',
    letterSpacing: '-0.32px',
  },
  media({ maxWidth: mobileWidth }, { width: '100%' })
);

export function Button(props: ButtonHTMLAttributes<HTMLButtonElement> & { 'data-testid'?: string }) {
  return <button {...props} className={buttonStyle} data-testid={props['data-testid'] || 'button'} />;
}
