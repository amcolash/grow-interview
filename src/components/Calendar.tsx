import React from 'react';
import { CaptionProps, DayPicker, useNavigation } from 'react-day-picker';
import 'react-day-picker/dist/style.css';
import { cssRule, style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface CalendarProps {
  date: Date;
  onChange: (date: Date) => void;
}

cssRule('.rdp', {
  fontFamily: 'Poppins',
  margin: 0,
});

export function Calendar(props: CalendarProps) {
  return (
    <DayPicker
      mode="single"
      selected={props.date}
      onDayClick={(day) => props.onChange(day)}
      disabled={{ from: new Date(), to: new Date(9999, 12, 31) }}
      formatters={{
        formatWeekdayName: (weekday) =>
          weekday.toLocaleDateString(undefined, {
            weekday: 'short',
          }),
      }}
      styles={{
        head_cell: {
          fontWeight: 'normal',
          color: 'var(--neutral-500, #A7AAAB)',
          fontSize: '12px',
          lineHeight: '18px',
        },
        cell: {
          fontSize: '14px',
          lineHeight: '22px',
        },
      }}
      components={{
        Caption: CustomCaption,
      }}
      modifiersStyles={{
        selected: {
          color: 'var(--brand-green-500, #025B4B)',
          background: 'var(--ivy-300, #B3CEC9)',
          fontWeight: '500',
        },
      }}
    />
  );
}

const captionStyle = style({
  display: 'flex',
  justifyContent: 'space-between',
  marginBottom: '32px',

  $nest: {
    '& button': {
      border: 'none',
      background: 'none',
      display: 'flex',
      padding: '0',
      alignItems: 'center',
      width: '24px',
      height: '24px',
      justifyContent: 'center',
      borderRadius: '100%',
    },
    '& button:hover': {
      background: 'var(--neutral-300, #e7eaeb)',
    },
    '& label': {
      fontSize: '16px',
      lineHeight: '24px',
      fontWeight: '500',
    },
  },
});

function CustomCaption(props: CaptionProps) {
  const { previousMonth, nextMonth, goToMonth, displayMonths } = useNavigation();
  const caption = props.displayMonth.toLocaleDateString(undefined, {
    month: 'long',
    year: 'numeric',
  });

  return (
    <div className={captionStyle}>
      <button
        onClick={() => {
          if (!previousMonth) return;
          goToMonth(previousMonth);
        }}
      >
        <ChevronLeft />
      </button>
      <label>{caption}</label>
      <button
        onClick={() => {
          if (!nextMonth) return;
          goToMonth(nextMonth);
        }}
        disabled={nextMonth ? nextMonth.getTime() > Date.now() : false}
      >
        <ChevronRight />
      </button>
    </div>
  );
}
