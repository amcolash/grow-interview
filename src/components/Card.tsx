import React from 'react';
import { style } from 'typestyle';

interface CardProps {
  children: React.ReactNode;
}

const cardListStyle = style({
  display: 'flex',
  flexDirection: 'column',
  gap: '20px',
  padding: '32px',

  borderRadius: '16px',
  background: '#FFF',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function Card(props: CardProps) {
  return <div className={cardListStyle}>{props.children}</div>;
}
