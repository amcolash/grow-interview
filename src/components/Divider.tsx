import React from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const dividerStyle = style(
  {
    borderLeft: '1px solid var(--neutral-300, #E7EAEB)',
    height: '64px',
    margin: '0 16px',
  },
  media({ maxWidth: mobileWidth }, { display: 'none' })
);

export function Divider() {
  return <div className={dividerStyle}></div>;
}
