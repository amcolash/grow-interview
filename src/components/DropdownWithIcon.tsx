import { Menu } from '@szhsin/react-menu';
import '@szhsin/react-menu/dist/transitions/slide.css';
import React, { FormEvent, MouseEventHandler } from 'react';
import { classes, media, style } from 'typestyle';

import { ReactComponent as ChevronDown } from '../icons/chevron-down.svg';
import { ReactComponent as ChevronUp } from '../icons/chevron-up.svg';

import { maxWidth } from '../consts';

interface DropdownWithIconProps {
  icon: React.ReactNode;
  label: string;
  value: React.ReactNode;
  children: React.ReactNode;
  onChange?: (value: any) => void;
  menuClassName?: string;
}

const wrapperStyle = style(
  {
    display: 'flex',
    padding: '12px',
    alignItems: 'center',
    gap: '24px',
    flex: '1 0 0',
    fontFamily: 'Poppins',
    borderRadius: '100px',
    userSelect: 'none',
    border: 'none',
    background: 'none',
    minWidth: '220px',

    $nest: {
      '&:hover, &:active': {
        background: 'var(--neutral-100, #F5F7F7)',
      },
      '.text': {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'flex-start',
        gap: '2px',
      },
      '.label': {
        color: 'var(--neutral-500, #A7AAAB)',
        fontSize: '12px',
        fontWeight: '500',
        lineHeight: '16px',
        letterSpacing: '0.6px',
        textTransform: 'uppercase',

        $nest: {
          label: {
            marginRight: '6px',
          },
        },
      },
      '.value': {
        color: 'var(--black, #000)',
        fontSize: '16px',
        lineHeight: '24px',
        letterSpacing: '-0.32px',
      },
    },
  },
  media({ maxWidth }, { padding: '2px' })
);

const menuStyle = style({
  display: 'flex',
  padding: '32px 16px',
  flexDirection: 'column',
  alignItems: 'flex-start',
  gap: '24px',
  borderRadius: '24px',
  background: 'var(--neutral-000, #FFF)',
  boxShadow: '0px 4px 24px 0px rgba(0, 0, 0, 0.12)',
  userSelect: 'none',
});

export function DropdownWithIcon(props: DropdownWithIconProps) {
  return (
    <Menu
      menuButton={({ open }) => <MenuButton {...props} open={open} />}
      onItemClick={(event) => {
        if (props.onChange) props.onChange(event.value);
      }}
      menuClassName={classes(menuStyle, props.menuClassName)}
      gap={10}
      transition
    >
      {props.children}
    </Menu>
  );
}

/** Internal Button Component */
interface MenuButtonProps {
  icon: React.ReactNode;
  label: string;
  value: React.ReactNode;
  open: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement> | undefined;
}

const MenuButton = React.forwardRef<HTMLButtonElement, MenuButtonProps>(function Internal(props, ref) {
  return (
    <button className={wrapperStyle} ref={ref} onClick={props.onClick}>
      {props.icon}

      <div className="text">
        <div className="label">
          <label>{props.label}</label>
          {props.open ? <ChevronUp /> : <ChevronDown />}
        </div>

        {props.value}
      </div>
    </button>
  );
});
