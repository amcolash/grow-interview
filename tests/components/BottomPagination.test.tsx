import { render } from '@testing-library/react';
import React from 'react';

import { BottomPagination } from '../../src/components/BottomPagination';

describe('BottomPagination', () => {
  it('renders on first page', () => {
    const { asFragment } = render(<BottomPagination currentPage={0} onPageChange={() => {}} totalPages={5} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders on third page', () => {
    const { asFragment } = render(<BottomPagination currentPage={2} onPageChange={() => {}} totalPages={5} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
