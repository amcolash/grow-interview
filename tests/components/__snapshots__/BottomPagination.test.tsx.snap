// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomPagination > renders on first page 1`] = `
<DocumentFragment>
  <ul
    aria-label="Pagination"
    class="f1l63s56"
    role="navigation"
  >
    <li
      class="previous disabled"
    >
      <a
        aria-disabled="true"
        aria-label="Previous page"
        class=" "
        rel="prev"
        role="button"
        tabindex="-1"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.2339 15C12.4444 15 12.6316 14.933 12.7719 14.799C13.076 14.531 13.076 14.062 12.7719 13.794L8.81871 9.99721L12.7719 6.22278C13.076 5.95477 13.076 5.48576 12.7719 5.21776C12.4912 4.92741 12 4.92741 11.7193 5.21776L7.22807 9.50586C6.92398 9.77387 6.92398 10.2429 7.22807 10.5109L11.7193 14.799C11.8596 14.933 12.0468 15 12.2339 15Z"
            fill="currentcolor"
          />
        </svg>
      </a>
    </li>
    <li
      class="selected"
    >
      <a
        aria-current="page"
        aria-label="Page 1 is your current page"
        rel="canonical"
        role="button"
        tabindex="-1"
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-label="Page 2"
        rel="next"
        role="button"
        tabindex="0"
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-label="Page 3"
        role="button"
        tabindex="0"
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-label="Page 4"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-label="Page 5"
        role="button"
        tabindex="0"
      >
        5
      </a>
    </li>
    <li
      class="next"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        class=""
        rel="next"
        role="button"
        tabindex="0"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.76608 15C7.55556 15 7.36842 14.933 7.22807 14.799C6.92398 14.531 6.92398 14.062 7.22807 13.794L11.1813 9.99721L7.22807 6.22278C6.92398 5.95477 6.92398 5.48576 7.22807 5.21776C7.50877 4.92741 8 4.92741 8.2807 5.21776L12.7719 9.50586C13.076 9.77387 13.076 10.2429 12.7719 10.5109L8.2807 14.799C8.14035 14.933 7.95322 15 7.76608 15Z"
            fill="currentcolor"
          />
        </svg>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`BottomPagination > renders on third page 1`] = `
<DocumentFragment>
  <ul
    aria-label="Pagination"
    class="f1l63s56"
    role="navigation"
  >
    <li
      class="previous"
    >
      <a
        aria-disabled="false"
        aria-label="Previous page"
        class=""
        rel="prev"
        role="button"
        tabindex="0"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.2339 15C12.4444 15 12.6316 14.933 12.7719 14.799C13.076 14.531 13.076 14.062 12.7719 13.794L8.81871 9.99721L12.7719 6.22278C13.076 5.95477 13.076 5.48576 12.7719 5.21776C12.4912 4.92741 12 4.92741 11.7193 5.21776L7.22807 9.50586C6.92398 9.77387 6.92398 10.2429 7.22807 10.5109L11.7193 14.799C11.8596 14.933 12.0468 15 12.2339 15Z"
            fill="currentcolor"
          />
        </svg>
      </a>
    </li>
    <li>
      <a
        aria-label="Page 1"
        role="button"
        tabindex="0"
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-label="Page 2"
        rel="prev"
        role="button"
        tabindex="0"
      >
        2
      </a>
    </li>
    <li
      class="selected"
    >
      <a
        aria-current="page"
        aria-label="Page 3 is your current page"
        rel="canonical"
        role="button"
        tabindex="-1"
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-label="Page 4"
        rel="next"
        role="button"
        tabindex="0"
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-label="Page 5"
        role="button"
        tabindex="0"
      >
        5
      </a>
    </li>
    <li
      class="next"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        class=""
        rel="next"
        role="button"
        tabindex="0"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.76608 15C7.55556 15 7.36842 14.933 7.22807 14.799C6.92398 14.531 6.92398 14.062 7.22807 13.794L11.1813 9.99721L7.22807 6.22278C6.92398 5.95477 6.92398 5.48576 7.22807 5.21776C7.50877 4.92741 8 4.92741 8.2807 5.21776L12.7719 9.50586C13.076 9.77387 13.076 10.2429 12.7719 10.5109L8.2807 14.799C8.14035 14.933 7.95322 15 7.76608 15Z"
            fill="currentcolor"
          />
        </svg>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
